WalletWasabi/WabiSabi/Client/RoundStatusUpdater.cs

```diff
@@ -0,0 +1,133 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class RoundStatusUpdater : PeriodicRunner
+	{
+		public RoundStatusUpdater(TimeSpan requestInterval, IWabiSabiApiRequestHandler arenaRequestHandler) : base(requestInterval)
+		{
+			ArenaRequestHandler = arenaRequestHandler;
+		}
+
+		private IWabiSabiApiRequestHandler ArenaRequestHandler { get; }
+		private Dictionary<uint256, RoundState> RoundStates { get; set; } = new();
+
+		private Dictionary<uint256, List<(TaskCompletionSource<RoundState> Task, Predicate<RoundState> Predicate)>> Awaiters { get; } = new();
+		private object AwaitersLock { get; } = new();
+
+		protected override async Task ActionAsync(CancellationToken cancellationToken)
+		{
+			var statusResponse = await ArenaRequestHandler.GetStatusAsync(cancellationToken).ConfigureAwait(false);
+			var responseRoundStates = statusResponse.ToDictionary(round => round.Id);
+
+			var updatedRoundStates = responseRoundStates.Where(round => RoundStates.ContainsKey(round.Key));
+			var newRoundStates = responseRoundStates.Where(round => !RoundStates.ContainsKey(round.Key));
+			var removedRoundStates = RoundStates.Where(round => !responseRoundStates.ContainsKey(round.Key));
+
+			var roundsToUpdate = updatedRoundStates.Where(updatedRound => RoundStates[updatedRound.Key] != updatedRound.Value)
+				.Union(newRoundStates)
+				.Union(removedRoundStates)
+				.Select(rs => rs.Key)
+				.ToList();
+
+			RoundStates = updatedRoundStates.Union(newRoundStates).ToDictionary(s => s.Key, s => s.Value);
+
+			lock (AwaitersLock)
+			{
+				// Handle round independent tasks.
+				if (Awaiters.TryGetValue(uint256.Zero, out var taskAndPredicateList))
```

Comment:
Maybe this would be useful:

```csharp
// Denotes that each round should be checked against a predicate list.
private static readonly uint256 WILDCARD_ROUND_ID = uint256.Zero;
```

Reply:
ACK

Reply:
That's an improvement. However, the problem is we have this special case when we should not, `roundId` is just a filtering condition as any other and for that reason it should be part of the predicate so, evaluate the predicate is all what we should do.

Reply:
Just to clarify. All what we need is a `list of predicate/cts` tuples. The special treatment that is given to the `roundId` is unnecessary.

Reply:
I was thinking about this. It would be much simpler BUT special treatment of roundId is required otherwise I could not be able to fail the predicate if the round is missing or removed.

Also there would be a minor performance hit (foreach all predicate on all roundStates).

Reply:
Take a look at this https://github.com/zkSNACKs/WalletWasabi/pull/5857